feat: Adding support for additional layer rendering options

kurtome opened the pull request:
# Description

This PR expands the functionality of `RenderableTileMap` to include support for more than just simple `TileLayer` layers. Adding support for:
 - `ImageLayer` rendering (note that `tiled.dart` was loading non-tile layers in the incorrect order until 0.8.4)
 - layer `parallaxX` / `parallaxY`
 - layer `opacity`
 - map `backgroundColor`

## tiled dependency

This PR relies on changes and bug fixes in `tiled.dart` 0.8.4 from this PR which has not been released yet: https://github.com/flame-engine/tiled.dart/pull/50

## Checklist


- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:` etc).
- [x] I have read the [Contributor Guide] and followed the process outlined for submitting PRs.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [x] I have updated/added relevant examples in `examples`.

## Breaking Change

<!-- Does your PR require Flame users to manually update their apps to accommodate your change? 

If the PR is a breaking change this should be indicated with suffix "!" 
(for example, `feat!:`, `fix!:`). See [Conventional Commit] for details.
-->

- [ ] Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

kurtome commented:
@spydon I tested this PR pretty extensively against my personal Flame project, and as far as I can tell it matches up quite well to how Tiled renders parallax, but I would love a second set of eyes or any suggestions for improvements

spydon commented on packages/flame_tiled/example/lib/main.dart:
```diff
@@ -38,4 +46,19 @@ class TiledGame extends FlameGame {
+    time += dt;
+    final tiledMap = mapComponent.tileMap.map;
+    if (time % 10 < 5) {
+      camera.moveTo(Vector2(
```
This (and the `else`) will create a lot of new unnecessary `Vector2` objects, you can create a vector outside of the update method that you use. (Since this will inspire how our users do this)

spydon commented on packages/flame_tiled/example/pubspec.yaml:
```diff
@@ -8,7 +8,8 @@ environment:
   flame: ^1.2.1
-  flame_tiled: ^1.6.0
+  flame_tiled:
+    path: "../../flame_tiled"
```
This should not use path, use `melos bootstrap`. Check `contributing.md` for how to set up melos.

spydon commented on packages/flame_tiled/lib/src/renderable_tile_map.dart:
```diff
@@ -194,11 +216,46 @@ class RenderableTiledMap {
+  /// Calculates the offset we need to apply to the canvas to compensate for
+  /// parallax positioning and scroll for the layer and the current camera position
+  /// https://doc.mapeditor.org/en/latest/manual/layers/#parallax-scrolling-factor
+  Vector2 calculateParallaxOffset(Camera camera, Layer layer) {
```
This should use global private vector objects that it re-uses instead of creating this many new vector objects each tick.

spydon commented on packages/flame_tiled/pubspec.yaml:
```diff
@@ -1,6 +1,6 @@
 name: flame_tiled
 description: Tiled support for the Flame game engine. This uses the tiled package and provides wrappers and components to be used inside Flame.
-version: 1.6.0
+version: 1.7.0
```
This should not be bumped in this PR.

spydon reviewed (commented):
Not sure what I think of passing the camera directly to the `RenderableTileMap` since that is something that isn't necesarilly used (especially now when `CameraComponent` will replace `Camera` in the long run)

kurtome commented:
@spydon Thanks for the feedback! I think I addressed the other issues, but this one:

> Not sure what I think of passing the camera directly to the RenderableTileMap since that is something that isn't necesarilly used (especially now when CameraComponent will replace Camera in the long run)

I agree it's not pretty and the tight coupling isn't ideal, honestly just not sure how else to implement it, 😬

kurtome commented:
@spydon I also had issues with the `map.tmx` file when opening it in Tiled and running the example app (tested running on Chrome), so the diff to that file basically recreates the entire look of the example app since basically all of the tiles weren't rendering for some reason in the original version.

spydon commented on packages/flame_tiled/lib/src/renderable_tile_map.dart:
```diff
@@ -192,17 +266,165 @@ class RenderableTiledMap {
+      opacity: layer.opacity,
+      alignment: Alignment.topLeft,
+      repeat: _repeat,
+      filterQuality: FilterQuality.medium,
```
Maybe this should be configurable?

kurtome replied:
I'm going to remove this line of code so it just uses the default value (`FilterQuality.low`), I think making this configurable would be a little to complicated for now since it's not clear if anyone needs it

spydon commented:
@ufrshubham if you have time, could you review this too? 🙂

ufrshubham commented on packages/flame_tiled/lib/src/renderable_tile_map.dart:
```diff
@@ -81,98 +103,150 @@ class RenderableTiledMap {
+  _RenderableTileLayer(
+    super.layer,
+    this._map,
+    this._camera,
```
If I understand correctly, the only time you need access to camera is in `render()`. So in that case, you can modify `TiledComponent` to have `HasGameRef` mixin (similar to `ParallaxComponent`). This will allow you to access the camera via `gameRef.camera`. You can then modify the render methods of `RenderableTiledMap` and `_RenderableLayer` to accept an additional Camera parameter as input.

kurtome replied:
That's a great point, would it be bad if it only worked for `FlameGame` and not `Game` though? (I think only `FlameGame` has `.camera`)

kurtome replied:
I think you're right, making it work for `FlameGame` is sufficient. 

The existing version of this PR shouldn't be breaking though since it's an optional parameter at the entry point of `TiledComponent`, but I'll refactor like you're suggesting sometime this week if I have time

kurtome replied:
I wound up setting defaulting the camera to `gameRef.camera` in `TiledComponent.gameResize` because it's needed during resize. This still allows client code to pass in their own camera if they want.

kurtome replied:
Ah, good point, I was moving too fast when I was refactoring and didn't notice that I can just use `canvasSize` directly. I'll make that change, which will make it so `ImageLayer` will work without a `Camera`, and therefore `Camera` will only be necessary for parallax support

ufrshubham commented on packages/flame_tiled/lib/src/tiled_component.dart:
```diff
@@ -10,7 +11,7 @@ import 'package:flame_tiled/src/renderable_tile_map.dart';
 /// Sprite Batches.
 /// {@endtemplate}
-class TiledComponent extends Component {
+class TiledComponent extends Component with HasGameRef {
```
Can you make it

```dart
class TiledComponent<T extends FlameGame> extends Component with HasGameRef<T> {
```

It will allow users to get their own gameRef of derived type.

kurtome replied:
done

ufrshubham commented:
LGTM. Also tested it with my platformer project😄, works great!
![simple-platformer-demo](https://user-images.githubusercontent.com/33748002/180154722-97ce6ee5-fa3b-4b2b-a715-8471c4737d71.gif)

spydon commented on packages/flame_tiled/lib/src/renderable_tile_map.dart:
```diff
@@ -192,17 +269,161 @@ class RenderableTiledMap {
+/// parallax positioning and scroll for the layer and the current camera
+/// position
+/// https://doc.mapeditor.org/en/latest/manual/layers/#parallax-scrolling-factor
+void _applyParallaxOffset(Canvas canvas, Camera camera, Layer layer) {
```
This is only used by one class right? I think it could be moved in to that class then.

kurtome replied:
It's used in two different layer classes, but I can put it in the `_RenderableLayer` base class 👍

kurtome replied:
Done

spydon commented on packages/flame_tiled/lib/src/renderable_tile_map.dart:
```diff
@@ -81,98 +108,148 @@ class RenderableTiledMap {
+  }
+}
+
+abstract class _RenderableLayer<T extends Layer> {
```
In the future (but that is for a much larger refactor) this should probably be a component, because then it can just have `PositionType.viewport` (or just added on the viewport with the new camera system) and it wouldn't have to have the camera sent in to it.

kurtome replied:
Agreed, I could do that refactor in a follow-up PR, but then there would be no way to use `RenderableTileMap` without using the Component system. Not sure if y'all want to support that

spydon reviewed (approved):
Good job!

kurtome commented:
Addressed some more comments and simplified `TiledComponent` so that is sets the `camera` in `onLoad` and just assumes you want to use the `FlameGame` camera since we're requiring that with `HasGameRef`
